src/lib.rs

```diff
@@ -0,0 +1,353 @@
+//! A Bech32m address encoder/decoder for PQC public keys.
+//!
+//! This crate supports:
+//! - Mainnet vs Testnet via HRP.
+//! - Disjoint byte ranges for version and pubkey type.
+//! - Support for multiple PQC public keys.
+//! - Full encode/decode round‑trip with Bech32m checksum per BIP‑350.
+
+use bech32::{Bech32m, Hrp, decode, encode};
+use hex::encode as hex_encode;
+use sha2::{Digest as ShaDigest, Sha256};
+use std::convert::TryFrom;
+use thiserror::Error;
+
+/// A Bech32 string is at most 90 characters long [BIP-173]
+const MAX_ADDRESS_LENGTH: usize = 90;
+
+/// Which network you’re on.
+///
+/// This determines the human‑readable part (HRP) of the Bech32m address.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Network {
+    /// Mainnet addresses use `yp…`
+    Mainnet,
+    /// Testnet addresses use `rh…`
+    Testnet,
+}
+
+impl Network {
+    /// Returns the Bech32m HRP for this network.
+    ///
+    /// # Panics
+    ///
+    /// This will panic if the hard-coded HRP literal fails to parse.
+    /// In practice `"yp"` and `"rh"` are spec-defined and known to be valid given `try_from`,
+    /// so this only ever fires if there’s a bug in this code or in `Hrp::parse`.
+    #[must_use]
+    pub fn hrp(self) -> Hrp {
+        match self {
+            Network::Mainnet => Hrp::parse("yp").expect("Mainnet HRP is valid"),
+            Network::Testnet => Hrp::parse("rh").expect("Testnet HRP is valid"),
+        }
+    }
+}
+
+/// Attempts to parse a network from a string.
+impl TryFrom<&str> for Network {
+    type Error = AddressDecodeError;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        match s {
+            "yp" => Ok(Network::Mainnet),
+            "rh" => Ok(Network::Testnet),
+            other => Err(AddressDecodeError::UnknownHrp(other.to_string())),
+        }
+    }
+}
+
+/// Address format version.
+///
+/// Codes are in the range `0x00..=0x3F` (64 total slots):
+/// - `0x00` = V1
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Version {
+    /// V1
+    V1,
+}
+
+impl Version {
+    /// Byte code to embed in the payload.
+    ///
+    /// Reserved range: 0x00..=0x3F (up to 64 versions).
+    #[must_use]
+    pub fn code(self) -> u8 {
+        match self {
+            Version::V1 => 0x00,
+        }
+    }
+
+    /// Reverse lookup from byte code to enum.
+    #[must_use]
+    pub fn from_code(code: u8) -> Option<Version> {
+        match code {
+            0x00 => Some(Version::V1),
+            _ => None,
+        }
+    }
+}
+
+/// Supported public-key types.
+///
+/// Codes are in the range `0x40..=0xFF` (192 total slots),
+/// giving us plenty of room for future PQC schemes:
+/// - `0x40` = ML-DSA 65
+/// - `0x41` = SLH-DSA SHA2 256 S
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum PubKeyType {
+    /// ML-DSA 65 public key.
+    MLDSA65,
+    /// SLH-DSA SHA2 256 S public key.
+    SLHDSASHA2_256S,
```

Comment:
I don't think we should actually have an SLH-DSA version yet until we've decided on the first parameter set we're going to support